### DISCUSSION TITLE
Help write a MicroPython library of ES8311 sensor.

### DISCUSSION THREAD

**Discussion by Vincent1-python** [micropython/micropython#15972] [Libraries & Drivers]
### Description

Help write a MicroPython library of ES8311 sensor.
I hope it is a library that uses Python completely, and can play sound and read sound by microphone, because the speaker and microphone chip of ESP32P4 is ES8311. thank you

### Code Size

As big as it can be, don't have too much flash space.

### Implementation

I hope the MicroPython maintainers or community will implement this feature, I intend to implement this feature and would submit a Pull Request if desirable

### Code of Conduct

Yes, I agree

**Comment by bidrohini** [comment]
You may see this for a reference: https://docs.firialabs.com/_modules/codec.html

**Comment by jonnor** [comment]
ES8311 is an audio codec (ADC+DAC) and uses I2S for audio data. You should be able to use [machine.I2S](https://docs.micropython.org/en/latest/library/machine.I2S.html) to communicate with it. There are many examples how to work with I2S in MicroPython at https://github.com/miketeachman/micropython-i2s-examples
You can also find some I2S example code in https://github.com/emlearn/emlearn-micropython/tree/master/examples/soundlevel_iir 

Note that the ES8311 codec may need some setup before it provides audio on the I2S bus of the format you want. This is done using I2C commands.

**Reply by Vincent1-python** [reply]
I just got stuck in the I2C command and asked for a library.

**Reply by jonnor** [reply]
Aha. I have not seen any I2C driver for this chip for MicroPython. I recommend looking at the code from a driver in another language, and porting the initialization sequence over. Here is one example, https://github.com/espressif/esp-bsp/tree/4dabefca0599b2249d356cb6a08268dd5e56bbaf/components/es8311
